noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+                    note_struct.name.0.contents
+                )),
+            })?;
+        // Look for the note struct implementation, generate a default one if it doesn't exist (in order to append methods to it)
+        let existing_impl = module.impls.iter_mut().find(|r#impl| match &r#impl.object_type.typ {
+            UnresolvedTypeData::Named(path, _, _) => path.last_segment().eq(&note_struct.name),
+            _ => false,
+        });
+        let note_impl = if let Some(note_impl) = existing_impl {
+            note_impl
+        } else {
+            let default_impl = TypeImpl {
+                object_type: trait_impl.object_type.clone(),
+                type_span: note_struct.name.span(),
+                generics: vec![],
+                methods: vec![],
+            };
+            module.impls.push(default_impl.clone());
+            module.impls.last_mut().unwrap()
+        };
+        // Identify the note type (struct name), its fields and its serialized length (generic param of NoteInterface trait impl)
+        let note_type = note_struct.name.0.contents.to_string();
+        let mut note_fields = vec![];
+        let note_serialized_len = match &trait_impl.trait_generics[0].typ {
+            UnresolvedTypeData::Named(path, _, _) => Ok(path.last_segment().0.contents.to_string()),
+            UnresolvedTypeData::Expression(UnresolvedTypeExpression::Constant(val, _)) => {
+                Ok(val.to_string())
+            }
+            _ => Err(AztecMacroError::CouldNotImplementNoteSerialization {
+                span: trait_impl.object_type.span,
+                secondary_message: Some(format!(
+                    "Cannot find note serialization length for: {}",
+                    note_type
+                )),
+            }),
+        }?;
+
+        // Automatically inject the header field if it's not present
+        if !note_struct.fields.iter().any(|(field_name, _)| field_name == "header") {
+            note_struct.fields.push((
+                ident("header"),
+                make_type(UnresolvedTypeData::Named(
+                    chained_dep!("aztec", "note", "note_header", "NoteHeader"),
+                    vec![],
+                    false,
+                )),
+            ));
+        }
+        for (field_ident, field_type) in note_struct.fields.iter() {
+            note_fields.push((
+                field_ident.0.contents.to_string(),
+                field_type.typ.to_string().replace("plain::", ""),
+            ));
+        }
+
+        let note_interface_impl_span = trait_impl.object_type.span;
+
+        // Generate the serialize_content method as
+        //
+        // fn serialize_content(self: {}) -> [Field; NOTE_SERIALIZED_LEN] {
+        //   [self.note_field1 as Field, self.note_field2.to_field()...]
+        // }
+        //
+        // It assumes every struct field can be converted either via the to_field() trait (structs) or cast as Field (integers)
```

Comment:
Should we also check if the field implements `Serializable` and recursively serialize? (in a future PR, for sure)

Reply:
That would be the job of the `derive(serlialize)` macro, but still would generate problems. What if a field is serialized to multiple `Field`s? We won't be able to properly reference them from the `NoteTypeFields` struct and our `FieldSelector`s would get mangled. This automatic system relies on the note fields occupying exactly 1 `Field` element in the serialization. Anything more advanced has to be manually implemented by the user...so I think we need an example of that.